flow-tests/test-root-context/src/test/java/com/vaadin/flow/uitest/ui/RefreshCurrentRouteIT.java

```diff
@@ -0,0 +1,100 @@
+
+public class RefreshCurrentRouteIT extends AbstractStreamResourceIT {
+
+    @Test
+    public void navigateToSameRoute_ensureSameInstanceAndCorrectEventCounts() {
+        open();
+
+        final String originalId = getString(ID);
+        final String originalLayoutId = getString(ROUTER_LAYOUT_ID);
+
+        assertInitialEventCounters();
+
+        $(NativeButtonElement.class).id(NAVIGATE_ID).click();
+
+        Assert.assertEquals(getString(ID), originalId);
+        Assert.assertEquals(getString(ROUTER_LAYOUT_ID), originalLayoutId);
+
+        // Nav events should have happened, attach/detach should not
+        Assert.assertEquals(1, getInt(ATTACHCOUNTER_ID));
+        Assert.assertEquals(0, getInt(DETACHCOUNTER_ID));
+        Assert.assertEquals(2, getInt(AFTERNAVCOUNTER_ID));
+        Assert.assertEquals(2, getInt(BEFOREENTERCOUNTER_ID));
+        Assert.assertEquals(1, getInt(BEFORELEAVECOUNTER_ID));
+    }
+
+    @Test
+    public void refreshCurrentRoute_ensureNewInstanceAndCorrectEventCounts_noNewLayout() {
+        open();
+
+        final String originalId = getString(ID);
+        final String originalLayoutId = getString(ROUTER_LAYOUT_ID);
+
+        assertInitialEventCounters();
+
+        $(NativeButtonElement.class).id(REFRESH_ID).click();
+
+        // UUID should be new since refresh creates new instance
+        Assert.assertNotEquals(getString(ID), originalId);
+        // Layout UUID should be same
+        Assert.assertEquals(getString(ROUTER_LAYOUT_ID), originalLayoutId);
+
+        // Event counters should equal original values
+        assertInitialEventCounters();
+    }
+
+    @Test
+    public void refreshCurrentRoute_ensureNewInstanceAndCorrectEventCounts_newLayout() {
+        open();
+
+        final String originalId = getString(ID);
+        final String originalLayoutId = getString(ROUTER_LAYOUT_ID);
+
+        assertInitialEventCounters();
+
+        $(NativeButtonElement.class).id(REFRESH_LAYOUTS_ID).click();
+
+        // UUID should be new since refresh creates new instance
+        Assert.assertNotEquals(getString(ID), originalId);
+        // UUID should be new since new layout instances were requested
+        Assert.assertNotEquals(getString(ROUTER_LAYOUT_ID), originalLayoutId);
+
+        // Event counters should equal original values
+        assertInitialEventCounters();
+    }
+
+    private void assertInitialEventCounters() {
+        Assert.assertEquals(1, getInt(ATTACHCOUNTER_ID));
+        Assert.assertEquals(0, getInt(DETACHCOUNTER_ID));
+        Assert.assertEquals(1, getInt(AFTERNAVCOUNTER_ID));
+        Assert.assertEquals(1, getInt(BEFOREENTERCOUNTER_ID));
+        Assert.assertEquals(0, getInt(BEFORELEAVECOUNTER_ID));
+    }
+
+    private String getString(String id) {
+        waitForElementPresent(By.id(id));
+        return findElement(By.id(id)).getText();
+    }
+
+    private int getInt(String id) {
+        return Integer.parseInt(getString(id));
+    }
+}
```

Comment:
nit: new line missing
```suggestion
}

```